include/partition/multi_level_graph.hpp

```diff
@@ -0,0 +1,191 @@
+                    Vector<typename SuperT::EdgeArrayEntry> edge_array_,
+                    Vector<EdgeOffset> node_to_edge_offset_)
+        : SuperT(std::move(node_array_), std::move(edge_array_)),
+          node_to_edge_offset(std::move(node_to_edge_offset_))
+    {
+    }
+
+    template <typename ContainerT>
+    MultiLevelGraph(const MultiLevelPartition &mlp,
+                    const std::uint32_t num_nodes,
+                    const ContainerT &edges)
+    {
+        auto highest_border_level = GetHighestBorderLevel(mlp, edges);
+        auto permutation = SortEdgesByHighestLevel(highest_border_level, edges);
+        auto sorted_edges_begin =
+            boost::make_permutation_iterator(edges.begin(), permutation.begin());
+        auto sorted_edges_end = boost::make_permutation_iterator(edges.begin(), permutation.end());
+        SuperT::InitializeFromSortedEdgeRange(num_nodes, sorted_edges_begin, sorted_edges_end);
+
+        // if the node ordering is sorting the border nodes first,
+        // the id of the maximum border node will be rather low
+        // enabling us to save some memory here
+        auto max_border_node_id = 0u;
+        for (auto edge_index : util::irange<std::size_t>(0, edges.size()))
+        {
+            if (highest_border_level[edge_index] > 0)
+            {
+                max_border_node_id =
+                    std::max(max_border_node_id,
+                             std::max(edges[edge_index].source, edges[edge_index].target));
+            }
+        }
+        BOOST_ASSERT(max_border_node_id < num_nodes);
+
+        auto edge_and_level_range = boost::combine(edges, highest_border_level);
+        auto sorted_edge_and_level_begin =
+            boost::make_permutation_iterator(edge_and_level_range.begin(), permutation.begin());
+        auto sorted_edge_and_level_end =
+            boost::make_permutation_iterator(edge_and_level_range.begin(), permutation.end());
+        InitializeOffsetsFromSortedEdges(
+            mlp, max_border_node_id, sorted_edge_and_level_begin, sorted_edge_and_level_end);
+    }
+
+    // Fast scan over all relevant border edges
+    auto GetBorderEdgeRange(const LevelID level, const NodeID node) const
+    {
+        auto begin = BeginBorderEdges(level, node);
+        auto end = SuperT::EndEdges(node);
+        return util::irange<EdgeID>(begin, end);
+    }
+
+    // Fast scan over all relevant internal edges, that is edges that will not
+    // leave the cell of that node at the given level
+    auto GetInternalEdgeRange(const LevelID level, const NodeID node) const
+    {
+        auto begin = SuperT::BeginEdges(node);
+        auto end = SuperT::BeginEdges(node) + node_to_edge_offset[node + level];
+        return util::irange<EdgeID>(begin, end);
+    }
+
+    EdgeID BeginBorderEdges(const LevelID level, const NodeID node) const
+    {
+        auto index = node * GetNumberOfLevels();
+        if (index >= node_to_edge_offset.size() - 1)
```

Comment:
`node_to_edge` will have the size defined by the maximum ID of border nodes. Right now that maximum ID is somewhere around `|V|` but if we implement a sorting of NodeIDs based on boundary level, we will save a lot of memory here (> 80%, typically less then 20% of the nodes are boundary nodes).